include/inexor/vulkan-renderer/frame_graph.hpp

```diff
@@ -0,0 +1,419 @@
+};
+
+class PhysicalImage : public PhysicalResource {
+    friend FrameGraph;
+
+private:
+    VkImage m_image{VK_NULL_HANDLE};
+    VkImageView m_image_view{VK_NULL_HANDLE};
+
+public:
+    PhysicalImage(VmaAllocator allocator, VkDevice device) : PhysicalResource(allocator, device) {}
+    PhysicalImage(const PhysicalImage &) = delete;
+    PhysicalImage(PhysicalImage &&) = delete;
+    ~PhysicalImage() override;
+
+    PhysicalImage &operator=(const PhysicalImage &) = delete;
+    PhysicalImage &operator=(PhysicalImage &&) = delete;
+};
+
+class PhysicalBackBuffer : public PhysicalResource {
+    friend FrameGraph;
+
+private:
+    const wrapper::Swapchain &m_swapchain;
+
+public:
+    PhysicalBackBuffer(VmaAllocator allocator, VkDevice device, const wrapper::Swapchain &swapchain)
+        : PhysicalResource(allocator, device), m_swapchain(swapchain) {}
+    PhysicalBackBuffer(const PhysicalBackBuffer &) = delete;
+    PhysicalBackBuffer(PhysicalBackBuffer &&) = delete;
+    ~PhysicalBackBuffer() override = default;
+
+    PhysicalBackBuffer &operator=(const PhysicalBackBuffer &) = delete;
+    PhysicalBackBuffer &operator=(PhysicalBackBuffer &&) = delete;
+};
+
+class PhysicalStage {
+    friend FrameGraph;
+
+private:
+    std::vector<wrapper::CommandBuffer> m_command_buffers;
+    VkDevice m_device;
+    VkPipeline m_pipeline{VK_NULL_HANDLE};
+    std::unique_ptr<wrapper::PipelineLayout> m_pipeline_layout;
+
+protected:
+    [[nodiscard]] VkDevice device() const {
+        return m_device;
+    }
+
+public:
+    explicit PhysicalStage(VkDevice device) : m_device(device) {}
+    PhysicalStage(const PhysicalStage &) = delete;
+    PhysicalStage(PhysicalStage &&) = delete;
+    virtual ~PhysicalStage();
+
+    PhysicalStage &operator=(const PhysicalStage &) = delete;
+    PhysicalStage &operator=(PhysicalStage &&) = delete;
+
+    /// @brief Retrieve the pipeline layout of this physical stage
+    // TODO(): This can be removed once descriptors are properly implemented in the frame graph
+    [[nodiscard]] VkPipelineLayout pipeline_layout() const {
+        return m_pipeline_layout->get();
+    }
+};
+
+class PhysicalGraphicsStage : public PhysicalStage {
+    friend FrameGraph;
+
+private:
+    VkRenderPass m_render_pass{VK_NULL_HANDLE};
+    std::vector<wrapper::Framebuffer> m_framebuffers;
+
+public:
+    explicit PhysicalGraphicsStage(VkDevice device) : PhysicalStage(device) {}
+    PhysicalGraphicsStage(const PhysicalGraphicsStage &) = delete;
+    PhysicalGraphicsStage(PhysicalGraphicsStage &&) = delete;
+    ~PhysicalGraphicsStage() override;
+
+    PhysicalGraphicsStage &operator=(const PhysicalGraphicsStage &) = delete;
+    PhysicalGraphicsStage &operator=(PhysicalGraphicsStage &&) = delete;
+};
+
+class FrameGraph {
+private:
+    VkDevice m_device;
+    VkCommandPool m_command_pool;
+    VmaAllocator m_allocator;
+    const wrapper::Swapchain &m_swapchain;
+    std::shared_ptr<spdlog::logger> m_log;
```

Comment:
```suggestion
    std::shared_ptr<spdlog::logger> m_log = spdlog::default_logger()->clone("frame-graph");
```
And remove from constructor.